config/module/tree.go

```diff
@@ -297,9 +301,145 @@ func (t *Tree) Load(s getter.Storage, mode GetMode) error {
 	// Set our tree up
 	t.children = children
 
+	// if we're the root module, we can now set the provider inheritance
+	if len(t.path) == 0 {
+		t.inheritProviderConfigs(nil)
+	}
+
 	return nil
 }
 
+// Once the tree is loaded, we can resolve all provider config inheritance.
+//
+// This moves the full responsibility of inheritance to the config loader,
+// simplifying locating provider configuration during graph evaluation.
+// The algorithm is much simpler now too. If there is a provider block without
+// a config, we look in the parent's Module block for a provider, and fetch
+// that provider's configuration. If that doesn't exist, we assume a default
+// empty config. Implicit providers can still inherit their config all the way
+// up from the root, so we walk up the tree and copy the first matching
+// provider into the module.
+func (t *Tree) inheritProviderConfigs(stack []*Tree) {
+	stack = append(stack, t)
+	for _, c := range t.children {
+		c.inheritProviderConfigs(stack)
+	}
```

Comment:
It took me a few skims to get what was going on here... I see now that we're sharing a backing array across all calls and then clobbering the last element of it as we iterate here. Eventually we have an array where the element right after our slice is the final child, but it's invisible because of where our slice is framed.

Assuming I got that right :grinning: I think it could help to note that here, since I think it's not immediately obvious that all of these calls are operating on the same backing array here.

Reply:
I should probably add a comment to prevent someone else form having to think it through ;) The slice here actually doesn't matter, because we only append so no recursion call will ever overwrite any values within our visible stack length. Because the current default to append to `nil` is a cap of 2, yes, we end up with a single child at the end, but it shouldn't matter. We could set a higher initial cap and see remnants of the longest and the final branch too if we wanted.